### PR TITLE
Router 7683 settle/refund logic

### LINKED ISSUE #4: Add logic related settle and refund int Router7683
Implement 
- [x] function _handle
- [x] function _handleSettlement
- [x] function _handleRefund
- [x] Happy path tests

### PR DESCRIPTION
**Pull request by lmcorbalan**
fixes #4 

TODO
- [x] Add assertIgpPayment to tests

### CODE REVIEW

File: solidity/src/Router7683.sol [hunk 1]
```diff
@@ -39,19 +42,34 @@ contract Router7683 is GasRouter, Base7683 {
...
+            }
+        } else {
+            for (uint i = 0; i < _orderIds.length; i++) {
+                _refundOrder(_orderIds[i], _origin);
+            }
+        }
+    }
+
+    function _handleSettlement(bytes32[] memory _orderIds, bytes32[] memory _receivers) internal virtual override {
+        _dispatchMessage(true, _orderIds, _receivers);
```
Comment (nambrot) [inline solidity/src/Router7683.sol:62]: Can't say I love this flag in the internal methods. IMO it should have explicit `_dispatchSettlement` or `dispatchRefund` methods and inside those you just call `_GasRouter_dispatch`. I know its a bit more duplication, but imo much more clear to read.

File: solidity/src/Router7683.sol [hunk 2]
```diff
@@ -61,4 +79,13 @@ contract Router7683 is GasRouter, Base7683 {
     function _localDomain() internal view override returns (uint32) {
         return localDomain;
     }
+
+    function _mustHaveSameOrigin(bytes32[] memory _orderIds) internal view returns (uint32 originDomain) {
```
Comment (nambrot) [inline solidity/src/Router7683.sol:83]: Is there a way we can avoid this gas expenditure? I.e. if an order that has a different origin domain is included in the error, can we just ignore on the origin chain? I'd assume that solvers will care about gas consumption

File: solidity/src/Base7683.sol [hunk 3]
```diff
@@ -57,7 +57,7 @@ abstract contract Base7683 is IOriginSettler, IDestinationSettler {
     // ============ Events ============
 
     event Filled(bytes32 orderId, bytes originData, bytes fillerData);
-    event Settle(bytes32[] orderIds, address[] receivers);
```
Comment (nambrot) [inline solidity/src/Base7683.sol:60]: I'm typically a stickler for keeping unrelated diffs in another PR, but i'll be lenient :P

File: solidity/src/Router7683.sol (hunk 2 above)
Comment (lmcorbalan) [reply solidity/src/Router7683.sol:83]: I think we could, but we need a way to determine where to send the message and I don't think we can get this from the orders. Maybe adding an explicit param to the `Base7386.settle` function?

Comment (lmcorbalan) [reply solidity/src/Router7683.sol:79]: I prefer not to add the para to the Base7386  `settle` and let the inheriting contract to handle it. Maybe some one wants to handle multi destinations settling.

Comment (nambrot) [reply solidity/src/Router7683.sol:79]: imo, its fine to just pick the first one, i think pushing this concern into the settlement-specific off-chain logic makes sense to me. Agree to not add this to base settle

**Review by nambrot — APPROVED** [review]

**Pull request merged**